cluster-autoscaler/cloudprovider/rancher/rancher_manager.go

```diff
@@ -0,0 +1,362 @@
+		return err
+	}
+
+	// cluster autoscaler specifies that we should wait until the node disappears from the underlying infrastructure
+	// before proceeding so we'll park here for a while
+	for retry := 0; retry < nodeDeletionVerifierRetryLimit; retry++ {
+		time.Sleep(nodeDeletionVerifierRetryDelay)
+
+		_, err := m.nodes.ByID(rancherNode.ID)
+		if err == nil {
+			continue
+		}
+
+		if clientbase.IsNotFound(err) {
+			// yay :)
+			return nil
+		}
+
+		klog.Warningf("unexpected error while checking node deletion (attempt %d/%d): %s",
+			retry, nodeDeletionVerifierRetryLimit, err)
+
+	}
+
+	return fmt.Errorf("unable to confirm node %s (%s) deletion for %s",
+		rancherNode.Name, rancherNode.ID, nodeDeletionVerifierRetryLimit*nodeDeletionVerifierRetryDelay)
+}
+
+// GetCachedNodePoolNodes returns a list of Nodes belonging to a NodePool
+func (m *RancherManager) GetCachedNodePoolNodes(id string) ([]*v3.Node, error) {
+	nodePoolNodes := make([]*v3.Node, 0)
+
+	for _, node := range m.nodeCacheByNodeName {
+		if node.NodePoolID == id {
+			nodePoolNodes = append(nodePoolNodes, node)
+		}
+	}
+
+	return nodePoolNodes, nil
+}
+
+// GetCachedNodeForKubernetesNode returns a Node object from Rancher given its Kubernetes' name
+func (m *RancherManager) GetCachedNodeForKubernetesNode(name string) (*v3.Node, error) {
+	node, exists := m.nodeCacheByNodeName[name]
+	if !exists {
+		return nil, fmt.Errorf("node  %s not found", name)
+	}
+
+	return node, nil
+}
+
+// GetNodeForKubernetesNode returns a Node object from Rancher given its Kubernetes' name
+func (m *RancherManager) GetNodeForKubernetesNode(nodeName string) (*v3.Node, error) {
+	listOpts := &types.ListOpts{
+		Filters: map[string]interface{}{
+			v3.NodeFieldClusterID: m.clusterId,
+			v3.NodeFieldNodeName:  nodeName,
+		},
+	}
+
+	nodes, err := m.nodes.List(listOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(nodes.Data) != 1 {
+		return nil, fmt.Errorf("node %s not found", nodeName)
+	}
+
+	return &nodes.Data[0], nil
+}
+
+// Cleanup disposes of resources (such as goroutines)
+func (m *RancherManager) Cleanup() error {
+	return nil
+}
+
+// Refresh invalidates the caches
+func (m *RancherManager) Refresh() error {
+	start := time.Now()
+	klog.Infof("Refreshing Rancher caches")
+
+	nodePools, err := m.fetchAllNodePoolsByID()
+	if err != nil {
+		return fmt.Errorf("failed to refresh node pool cache: %s", err)
+	}
+
+	nodesByNodeName, err := m.fetchAllNodesByNodeName()
+	if err != nil {
+		return fmt.Errorf("failed to refresh node cache: %s", err)
+	}
+
+	m.nodePoolCacheByID = nodePools
+	m.nodeCacheByNodeName = nodesByNodeName
+
+	elapsed := time.Since(start)
+	klog.Infof("Refresh completed in %s. There are currently %d node pools and %d nodes.",
+		elapsed, len(m.nodePoolCacheByID), len(m.nodeCacheByNodeName))
+
+	return nil
+}
+
+func hasAutoScalingEnabled(nodePool *v3.NodePool) bool {
+	return nodePool.Annotations[autoScalingEnabledAnnotation] == "true"
+}
+
+// caching stuff
```

Comment:
I believe this comment should be before `Refresh()` declaration

Reply:
Those are the actual helpers added only for caching. We didn't need to fetch all nodes / node pools before since we requested them on demand